auth/oidc.go

```diff
@@ -399,8 +403,36 @@ func (op *OIDCProvider) standardDiscovery(discoveryURL string) (metadata Provide
 	return metadata, verifier, err
 }
 
-func getOIDCUsername(provider *OIDCProvider, subject string) string {
-	return fmt.Sprintf("%s_%s", provider.UserPrefix, url.QueryEscape(subject))
+// getOIDCUsername returns the username to be used as the Sync Gateway username.
+func getOIDCUsername(provider *OIDCProvider, identity *Identity) (username string, err error) {
+	if provider.UsernameClaim != "" {
+		value, ok := identity.Claims[provider.UsernameClaim]
+		if !ok {
+			return "", fmt.Errorf("oidc: specified claim %q not found in id_token, identity: %v", provider.UsernameClaim, identity)
+		}
+		if username, err = formatUsername(value); err != nil {
+			return "", err
+		}
+		if provider.UserPrefix == "" {
+			return url.QueryEscape(username), nil
+		}
+		return fmt.Sprintf("%s_%s", provider.UserPrefix, url.QueryEscape(username)), nil
+	}
+	return fmt.Sprintf("%s_%s", provider.UserPrefix, url.QueryEscape(identity.Subject)), nil
+}
+
+// formatUsername returns the string representation of the given username value.
+func formatUsername(value interface{}) (string, error) {
+	switch valueType := value.(type) {
+	case string:
+		return valueType, nil
+	case json.Number:
+		return string(valueType), nil
+	case float64:
+		return strconv.FormatFloat(valueType, 'e', -1, 64), nil
```

Comment:
This code isn't going to be currently used, as callers are (correctly) using UseNumber, so we could potentially remove this case.  If we're going to keep it, wouldn't it be preferable to use format 'g', to better preserve format for non-large exponents?